aws-cpp-sdk-transfer/source/transfer/TransferManager.cpp

```diff
@@ -65,6 +65,53 @@ namespace Aws
             }
         }
 
+        Transfer::TransferStatus TransferManager::WaitUntilAllFinished(int64_t timeoutMs)
+        {
+            Transfer::TransferStatus status = Transfer::TransferStatus::IN_PROGRESS;
+            do
+            {
+                std::unique_lock<std::mutex> lock(m_tasksMutex);
+                if (m_tasks.empty())
+                {
+                    status = Transfer::TransferStatus::COMPLETED;
+                    break;
+                }
+                const auto waitStarted = std::chrono::steady_clock::now();
+                m_tasksSignal.wait_for(lock, std::chrono::milliseconds(timeoutMs));
+                timeoutMs -= std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - waitStarted).count();
+            }
+            while(status != Transfer::TransferStatus::COMPLETED && timeoutMs > 0);
+
+            return status;
+        }
+
+        void TransferManager::CancelAll()
+        {
+            Aws::UnorderedSet<std::shared_ptr<TransferHandle>> tasksCopy;
```

Comment:
cancelling on a copy because `it->Cancel();` will Submit more tasks into the queue, potentially leading into a deadlock.